### PR TITLE
Implement cookie consent

### PR DESCRIPTION
**Pull request by tsteur**
refs https://github.com/matomo-org/matomo/issues/15948

Works the same way as tracking consent. However, you need to change `Consent` to `CookieConsent`:

* `_paq.push(['requireConsent']);` --> `_paq.push(['requireCookieConsent']);`
* `_paq.push(['setConsentGiven']);` --> `_paq.push(['setCookieConsentGiven']);`
* `_paq.push(['rememberConsentGiven']);` --> `_paq.push(['rememberCookieConsentGiven']);`
* `_paq.push(['forgetConsentGiven']);` --> `_paq.push(['forgetCookieConsentGiven']);`

Renamed `enableCookies` for consistency to `setCookieConsentGiven`. The method was only added last week so that's not an issue.

`disableCookies` still exists and won't change meaning it will always disable cookies no matter if consent was given or not.

Basically it works the very same way as regular consent see https://developer.matomo.org/guides/tracking-javascript-guide#asking-for-consent

Eg you add `_paq.push(['requireCookieConsent']);` to your tracking code and it will keep on tracking but won't set any cookies.

Unless you call either `_paq.push(['setCookieConsentGiven'])` on every page view or once `_paq.push(['rememberCookieConsentGiven']);` to store the consent in a cookie.

Once consent was revoked, you can call `_paq.push(['forgetCookieConsentGiven']);` to remove the existing cookie.

documentation in https://github.com/matomo-org/developer-documentation/pull/359/files

Failing tests aren't related to this PR

### CODE REVIEW

**Comment by tsteur** [comment]
build js

**Comment by tsteur** [comment]
fyi @Findus23

File: js/piwik.js [hunk 1]
```diff
@@ -6828,25 +6834,83 @@ if (typeof window.Piwik !== 'object') {
...
+             * or call `_paq.push(['setCookieConsentGiven'])` on each page view at any time after calling `requireCookieConsent`.
+             *
+             * When the user gives you the consent to set cookies, you can also call `_paq.push(['rememberCookieConsentGiven', optionalTimeoutInHours])`
+             * and for the duration while the cookie consent is remembered, any call to `requireCoookieConsent` will be automatically ignored
+             * until you call `forgetCookieConsentGiven`.
+             * `forgetCookieConsentGiven` needs to be called when the user removes consent for using cookies. This means if you call `rememberCookieConsentGiven` at the
+             * time the user gives you consent, you do not need to ever call `_paq.push(['setCookieConsentGiven'])` as the consent
+             * will be detected automatically through cookies.
              */
-            this.enableCookies = function () {
```
Comment (tsteur) [inline js/piwik.js:6833]: the method enableCookies was twice in js/piwik.js because of some merge I suppose.

File: js/piwik.js [hunk 2]
```diff
@@ -7754,6 +7814,7 @@ if (typeof window.Piwik !== 'object') {
 
                 deleteCookie(CONSENT_COOKIE_NAME, configCookiePath, configCookieDomain);
                 setCookie(CONSENT_REMOVED_COOKIE_NAME, new Date().getTime(), thirtyYears, configCookiePath, configCookieDomain, configCookieIsSecure);
+                this.forgetCookieConsentGiven();
```
[the remark this answers is not in the document]
Comment (diosmosis) [reply js/piwik.js:7817]: The interaction between tracking consent and cookie consent is confusing to me. Are users expected to use both always? Or sometimes just tracking consent? They seem like they could be completely separate, but some of the tracking consent methods affect cookie consent. Should they be kept completely separate?

[the remark this answers is not in the document]
Comment (tsteur) [reply js/piwik.js:7817]: @diosmosis it's either - or. You only use one of them. 

If you remove tracking consent, then you indirectly also remove the cookie consent since you basically don't want to be tracked anymore no cookies should be set anymore.

So people use either cookie consent or tracking consent. 

Tracking consent basically includes cookie consent and when removing consent, we're no longer allowed to set cookies. The cookies are then also no longer needed since we no longer track them anyway.

Generally, if you want more accurate data and get exact unique visitors, new/returning visitors etc then you need cookies. Meaning you need to ask for cookie consent. We still track the user in any case. But only when consent is given we track with cookies, otherwise without.

If you're also tracking personal data (full IP, user ID, etc) then you need to ask for tracking consent. If consent is not given, we're not allowed to track at all thus we're also not allowed to set cookies. If consent is given, we can use cookies.

Does this help?

[the remark this answers is not in the document]
Comment (diosmosis) [reply js/piwik.js:7817]: So the two main use cases are "require user to opt in to all tracking (which implies allowing cookies)" and "track user anyway, but require user to opt in to storing cookies for more accurate tracking"?

It might be useful to have documentation for "how to build a consent form" or something eventually.

[the remark this answers is not in the document]
Comment (tsteur) [reply js/piwik.js:7817]: We have it explained in https://developer.matomo.org/guides/tracking-javascript-guide#asking-for-consent and in Matomo itself in `Admin -> Privacy -> Consent`. Created a PR for adjusting the docs in https://github.com/matomo-org/developer-documentation/pull/359/files and will also adjust the docs in Matomo directly after the PR is merged.

We'll also be creating a new FAQ as part of https://github.com/matomo-org/matomo/issues/15948 . It's mostly interesting for consent managers but also for people who build their own consent popup.

[the remark this answers is not in the document]
Comment (tsteur) [reply js/piwik.js:7817]: Yes those are the two main use cases @diosmosis
